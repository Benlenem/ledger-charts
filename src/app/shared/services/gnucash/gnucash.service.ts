import { Injectable } from '@angular/core';
import Decimal from 'decimal.js';
import * as moment from 'moment';
import { Observable } from 'rxjs';
import { Transaction } from '../../models/transaction';
import { xmlToJson } from '../../utils/utils';

const AmountRegex = /(.*)\/(.*)/;

@Injectable()
export class GnucashService {
  constructor() {}

  constructAccountName = (id, map) => {
    const acc = map[id];

    if (!acc) {
      return '';
    }

    const name = acc.name;
    const parent = acc.parent;

    if (!parent) {
      return name;
    }

    const parentFullName = this.constructAccountName(parent, map);
    return parentFullName ? parentFullName + ':' + name : name;
  }

  parseGnucashString(gnucashString: string): Observable<Transaction[]> {
    return new Observable(obs => {
      const jsGnucash = xmlToJson(new DOMParser().parseFromString(gnucashString, 'text/xml'));

      const jsAcc = jsGnucash['gnc-v2']['gnc:book']['gnc:account'];

      const accounts = jsAcc
        .map(a => ({
          name: a['act:name']['#text'],
          id: a['act:id']['#text'],
          parent: a['act:parent'] ? a['act:parent']['#text'] : undefined,
        }))
        .filter(a => a.name !== 'Root Account')
        .reduce((map, account) => {
          map[account.id] = account;
          return map;
        }, {});

      const toto = Object.keys(accounts).reduce((map, account) => {
        map[account] = this.constructAccountName(account, accounts);
        return map;
      }, {});

      let jsTr = jsGnucash['gnc-v2']['gnc:book']['gnc:transaction'];

      if (!Array.isArray(jsTr)) {
        jsTr = [jsTr];
      }

      const transactions: Transaction[] = jsTr.map(t => {
        const currency = t['trn:currency']['cmdty:id']['#text'];
        const datePosted = t['trn:date-posted']['ts:date']['#text'];
        const descriptionNode = t['trn:description']['#text'];
        const description = descriptionNode ? descriptionNode.replace(/(\r\n|\n|\r| {2,})/gm, '') : '';

        let splits = t['trn:splits']['trn:split'];

        if (!Array.isArray(splits)) {
          splits = [splits];
        }

        return {
          header: {
            date: moment.utc(datePosted, 'YYYY-MM-DD HH-mm-ss Z'),
            title: description,
          },
          postings: splits.map(s => {
            const rawAmount = s['split:value']['#text'];
            const reg = AmountRegex.exec(rawAmount);
            const memo = s['split:memo'] ? s['split:memo']['#text'] : undefined;
            const accountId = s['split:account']['#text'];
            const fullAccountName = toto[accountId];
            return {
              currency,
              account: fullAccountName || accountId,
              amount: reg ? new Decimal(reg[1]).dividedBy(reg[2]) : new Decimal(0),
              comment: memo,
            };
          }),
        };
      });

      obs.next(transactions || []);
      obs.complete();
    });
  }
}
