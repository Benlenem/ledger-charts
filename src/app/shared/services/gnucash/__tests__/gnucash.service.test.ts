import {
  emptyFile,
  fileWithOneTransaction,
  fileWithTwoTransactions,
  fileWithOneSplit,
  fileWithNoAMount,
  fileWithLineBreak,
} from './fixtures';
import { GnucashService } from '../gnucash.service';
import 'rxjs/add/operator/toPromise';

describe(GnucashService.name, () => {
  it('parses empty file', () => {
    const service = new GnucashService();

    return expect(service.parseGnucashString(emptyFile).toPromise()).rejects.toMatchSnapshot();
  });

  it('parses file with one transaction', () => {
    const service = new GnucashService();

    return expect(service.parseGnucashString(fileWithOneTransaction).toPromise()).resolves.toMatchSnapshot();
  });

  it('parses file with two transactions', () => {
    const service = new GnucashService();

    return expect(service.parseGnucashString(fileWithTwoTransactions).toPromise()).resolves.toMatchSnapshot();
  });

  it('parses file with only one split', () => {
    const service = new GnucashService();

    return expect(service.parseGnucashString(fileWithOneSplit).toPromise()).resolves.toMatchSnapshot();
  });

  it('parses file with missing amount', () => {
    const service = new GnucashService();

    return expect(service.parseGnucashString(fileWithNoAMount).toPromise()).resolves.toMatchSnapshot();
  });

  it('parses file with line break', () => {
    const service = new GnucashService();

    return expect(service.parseGnucashString(fileWithLineBreak).toPromise()).resolves.toMatchSnapshot();
  });
});
