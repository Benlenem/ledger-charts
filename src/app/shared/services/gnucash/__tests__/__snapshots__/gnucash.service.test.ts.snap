// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GnucashService parses empty file 1`] = `[TypeError: Cannot read property 'gnc:book' of undefined]`;

exports[`GnucashService parses file with line break 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2016-05-27T10:59:00.000Z",
      "title": "VIREMENT EMIS - renflouement fin mai",
    },
    "postings": Array [
      Object {
        "account": "Actif",
        "amount": "100",
        "comment": undefined,
        "currency": "EUR",
      },
      Object {
        "account": "Revenus",
        "amount": "-100",
        "comment": "WEB ORSINI BENJAMIN OU M renflou",
        "currency": "EUR",
      },
    ],
  },
]
`;

exports[`GnucashService parses file with missing amount 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2016-05-27T10:59:00.000Z",
      "title": "VIREMENT EMIS - renflouement fin mai",
    },
    "postings": Array [
      Object {
        "account": "Actif",
        "amount": "0",
        "comment": undefined,
        "currency": "EUR",
      },
    ],
  },
]
`;

exports[`GnucashService parses file with one transaction 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2016-05-27T10:59:00.000Z",
      "title": "VIREMENT EMIS - renflouement fin mai",
    },
    "postings": Array [
      Object {
        "account": "Actif",
        "amount": "100",
        "comment": undefined,
        "currency": "EUR",
      },
      Object {
        "account": "Revenus",
        "amount": "-100",
        "comment": "WEB ORSINI BENJAMIN OU M renflou",
        "currency": "EUR",
      },
    ],
  },
]
`;

exports[`GnucashService parses file with only one split 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2016-05-27T10:59:00.000Z",
      "title": "VIREMENT EMIS - renflouement fin mai",
    },
    "postings": Array [
      Object {
        "account": "Actif",
        "amount": "100",
        "comment": undefined,
        "currency": "EUR",
      },
    ],
  },
]
`;

exports[`GnucashService parses file with two transactions 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2016-05-27T10:59:00.000Z",
      "title": "VIREMENT EMIS - renflouement fin mai",
    },
    "postings": Array [
      Object {
        "account": "Actif",
        "amount": "100",
        "comment": undefined,
        "currency": "EUR",
      },
      Object {
        "account": "Revenus",
        "amount": "-100",
        "comment": "WEB ORSINI BENJAMIN OU M renflou",
        "currency": "EUR",
      },
    ],
  },
  Object {
    "header": Object {
      "date": "2016-05-30T10:59:00.000Z",
      "title": "LEROY MERLIN MERIGNAC 27/05",
    },
    "postings": Array [
      Object {
        "account": "Actif",
        "amount": "100.8",
        "comment": undefined,
        "currency": "EUR",
      },
      Object {
        "account": "Revenus",
        "amount": "-100.8",
        "comment": "PAIEMENT PAR CARTE",
        "currency": "EUR",
      },
    ],
  },
]
`;
